Add dropdown version of Filter by Stock Status

kmanijak opened the pull request:
**Main changes:**

- add dropdown version to the Filter by Stock Status
- add option for `multiple` or `single` choice to align with Filter by Attribute

**Minor changes:**
- alignment between Filter by Attribute and Stock:
  - Editor settings options: wording and settings order

Know issue:
- [minor] When typing in the dropdown input, when you put whitespace, there's one character misalignment between what's input and what is underlined. This is due to the limitations of the `FormTokenField` API. It could be fixed, but requires amendments to the filter keys, which I believe is too complex comparing to the gain. But I'm open for suggestions.
<img width="265" alt="image" src="https://user-images.githubusercontent.com/20098064/206134383-b07eacee-5007-44c1-8c5c-6077f073f207.png">
_Input: "In stoc", Underline "In st", missing two characters_


Fixes #7770 

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

### Screenshots
Watch the video of Dropdown component usage:

https://user-images.githubusercontent.com/20098064/206128900-2cdb4bfe-c131-48d1-b395-8fa3fc07dafc.mov

| Case | Before | After |
| ------ | ----- | ----- |
|   Editor Settings     |    <img width="282" alt="Screen Shot 2022-12-07 at 09 38 58 AM" src="https://user-images.githubusercontent.com/20098064/206129793-adbe0756-f717-4fa8-baf0-d5f36dd327cd.png">   |   <img width="284" alt="Screen Shot 2022-12-07 at 09 37 11 AM" src="https://user-images.githubusercontent.com/20098064/206129728-f239e2f6-495e-45c7-bea1-305bd01a247b.png">   |

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [x] E2E tests

#### User Facing Testing

Check out [the video above](
https://user-images.githubusercontent.com/20098064/206128900-2cdb4bfe-c131-48d1-b395-8fa3fc07dafc.mov) which will help you go through the testing instructions

Prerequisites:
1. Have at least few products with different stock status: in stock, out of stock, on backorder
2. Go to Editor
3. Add Filters pattern
4. Add All Products

##### Dropdown component - single
1. Change settings of Filter by Stock to: single / dropdown
2. Open dropdown
  - **Expected**: dropdown has stock status options to choose from
3. Choose one option
  - **Expected**: page refreshes, filter is applied: visible in Active Filters and Products are filtered out
4. Choose other option
  - **Expected**: previous filter is replaced by the new one, page refreshes, filter is applied: visible in Active Filters and Products are filtered out

##### Dropdown component - multiple
1. Change settings of Filter by Stock to: multiple / dropdown
2. Open dropdown
  - **Expected**: dropdown has stock status options to choose from
3. Choose one option
  - **Expected**: page refreshes, filter is applied: visible in Active Filters and Products are filtered out
4. Choose other option
  - **Expected**: new filter is added to the new one, page refreshes, both filters ("or") is applied: visible in Active Filters and Products are filtered out

##### List component
1. Change settings of Filter by Stock to: list
2. Repeat the above for both: single and multiple

##### Apply button
1. In the Editor choose multiple Dropdown again and enable "Show 'Apply filters' button" option
2. In the frontend choose stock filters
  - **Expected**: filters are not applied until "Apply" button is clicked. Once clicked page refreshes and filters are applied

##### Other scenarios
Try to break it! 😉 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Filter by Stock: Add dropdown version.
> Filter by Stock: Add option to choose multiple or single option.

kmanijak committed:
Extend Filter by Stock Editor options with dropdown and single/multiple choice

kmanijak committed:
Add dropdown implementation for Filter by Stock Status

kmanijak committed:
Merge branch 'trunk' into add/filter-by-stock-dropdown

kmanijak committed:
Adjust font-sizes to the rest of the filters

kmanijak committed:
Fix logical bug in onChange handler of Filter by Stock

kmanijak committed:
Add tests to Filter by Stock: dropdown and list variants

Change test file extension from .js to .tsx, so it handles types as well

kmanijak committed:
Add E2E test to Filter by Stock checking if display style can be toggled

kmanijak committed:
When typing in Filter by Stock dropdown, handle the space so it highlights the suggestions

kmanijak committed:
Revert Editor Settings label to align to Filter by Rating and Attribute

kmanijak committed:
Typescript errors fixes and adopting previous Filter by Stock tests to new setup

sunyatasattva commented:
Could you clarify which limitations you mean on the `FormTokenField`? I've used the component from Gutenberg in the exact same scenario and it works fine in my use-case:

<img width="267" alt="Screenshot 2022-12-07 at 23 50 30" src="https://user-images.githubusercontent.com/1847066/206313812-dfee9ca0-e495-4fab-964e-f2841d9790c7.png">

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -169,15 +185,19 @@ const StockStatusFilterBlock = ( {
 					label: (
 						<Label
-							name={ decodeEntities( status.name ) }
+							name={ decodeEntities( status.name ) as string }
```
I probably wouldn't cast this here. It's just swiping under the rug the real problem: which is the missing type in `STOCK_STATUS_OPTIONS`. You don't have to fix that if you believe it is outside of the scope of this PR (the real fix could bring you deep into a rabbit hole; though an ok fix could require just a few minutes probably), but in that case I wouldn't hide the problem by casting this as `string`.

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -221,6 +241,8 @@ const StockStatusFilterBlock = ( {
 		changeUrl( newUrl );
 	};
 
+	const multiple = blockAttributes.selectType !== 'single';
```
Super nit-pick here and feel free to disagree with me on this and ignore it. I am not a big fan of the name `multiple`, as it doesn't say much. I like names which are semantically appropriate. A better name could be `allow(s)MultipleOptions` or just `allow(s)Multiple`.

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+								'Select stock status',
+								'woo-gutenberg-products-block'
+							) }
+							onChange={ ( tokens: string[] ) => {
```
I think `: string[]` is unnecessary, TS will do the correct type inference here.

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+							} ) }
+							style={ {
+								...borderProps.style,
+								borderStyle: 'none',
```
Why is this inline style necessary?

kmanijak replied:
It was not, I was reusing some of the Filter by Attribute work and this is a leftover, thanks!

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+							) }
+							onChange={ ( tokens: string[] ) => {
+								if ( ! multiple && tokens.length > 1 ) {
+									tokens = [ tokens[ tokens.length - 1 ] ];
```
What about this instead?
```suggestion
									tokens = tokens.slice(-1);
```

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+
+								const added = difference( tokens, checked );
+
+								if ( added.length === 1 ) {
```
I'm not 100% sure I understand. What happens if `.length` is > 1? Or that can never happen because it is assumed that `onChange` is called at every selection? What happens if someone copy+pastes a string with multiple valid tokens into the input?

kmanijak replied:
> What happens if .length is > 1?

To be honest, I was trying to find a case when the `.length` is > 1 and couldn't find it. But that check (`if ( added.length === 1 )` ) is there in Filter by Attribute and I assumed it's there for a reason and there may be a case I'm not aware of.

> Or that can never happen because it is assumed that onChange is called at every selection?

I think that's actually the point. The only thing I could think of was that the rest of the logic is not prepared for more than one token difference at once (e.g. screen reader communicates a single token was added)

> What happens if someone copy+pastes a string with multiple valid tokens into the input?

User has to explicitly confirm the selection of each token, so they can only copy-paste the search string, not the token itself. Hence, I think it's impossible to add more than single token at once. Unless you know such case? I'm open for suggestions here.

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+								}
+							} }
+							value={ checked }
+							displayTransform={ ( value: string ) => {
```
Also unnecessary type here.

kmanijak replied:
As in the other comment:
Value has implicitly type any 🤔 or do you mean to explicitly apply type any?

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+								return result ? result.textLabel : value;
+							} }
+							saveTransform={ formatSlug }
+							messages={ {
```
Just curious: there is `removed` and `remove` but `added` without `add`, is that intentional? Does `messages` not support that?

kmanijak replied:
That's right, messages are asymmetric. There's no `add` messages even in the newest version ([docs](https://developer.wordpress.org/block-editor/reference-guides/components/form-token-field/))

sunyatasattva commented on assets/js/blocks/stock-filter/block.tsx:
```diff
@@ -385,18 +418,104 @@ const StockStatusFilterBlock = ( {
+								const removed = difference( checked, tokens );
+								if ( removed.length === 1 ) {
+									onChange( removed[ 0 ] );
+								}
```
This function is pretty long to be inlined in the element TBH, I'd refactor it in its own separate function.

sunyatasattva commented on assets/js/blocks/stock-filter/edit.tsx:
```diff
@@ -51,10 +65,62 @@ const Edit = ( {
+							'woo-gutenberg-products-block'
+						) }
+						value={ selectType || 'multiple' }
+						onChange={ ( value: string ) =>
```
Unnecessary type.

kmanijak replied:
Value has implicitly type `any` 🤔 or do you mean to explicitly apply type `any`?

sunyatasattva commented on assets/js/blocks/stock-filter/edit.tsx:
```diff
@@ -51,10 +65,62 @@ const Edit = ( {
+							'woo-gutenberg-products-block'
+						) }
+						value={ selectType || 'multiple' }
+						onChange={ ( value: string ) =>
+							setAttributes( {
+								selectType: value,
+							} )
+						}
```
You could make this slightly more concise like so:

```suggestion
						onChange={ ( selectType ) => setAttributes( { selectType } ) }
```

Not required, but just a suggestion.

kmanijak replied:
Unfortunately `selectType` is already declared in the scope (`attributes` destructuring). I could apply the suggestion, but that means making other parts a bit less concise, so I'd stick to current version.

sunyatasattva commented on assets/js/blocks/stock-filter/edit.tsx:
```diff
@@ -51,10 +65,62 @@ const Edit = ( {
+						}
+						className="wc-block-attribute-filter__multiple-toggle"
+					>
+						<ToggleGroupControlOption
+							value="multiple"
+							label={ __(
+								'Multiple',
+								'woo-gutenberg-products-block'
+							) }
+						/>
+						<ToggleGroupControlOption
+							value="single"
+							label={ __(
+								'Single',
+								'woo-gutenberg-products-block'
+							) }
+						/>
```
You know what? Thinking of this, this should just be a toggle switch, I believe. It's basically “Allow selecting multiple options?”: yes / no.

kmanijak replied:
Good catch and I totally agree! I took it from Filter to keep them aligned. Having said that and keeping in mind I did the same for Filter by Rating as well ([PR](https://github.com/woocommerce/woocommerce-blocks/pull/7771)), do you mind if I create an issue and change the settings in all three places at once, once both Filters (by Rating and by Stock Status) are merged?

sunyatasattva commented on assets/js/blocks/stock-filter/edit.tsx:
```diff
@@ -51,10 +65,62 @@ const Edit = ( {
+							'woo-gutenberg-products-block'
+						) }
+						value={ displayStyle }
+						onChange={ ( value: string ) =>
+							setAttributes( {
+								displayStyle: value,
+							} )
+						}
```
Same comments as above.

kmanijak replied:
Unfortunately `displayStyle` is already declared in the scope (from `attributes` destructuring). I could apply the suggestion, but that means making other parts a bit less concise, so I'd stick to current version.

sunyatasattva commented on assets/js/blocks/stock-filter/preview.tsx:
```diff
@@ -8,15 +8,18 @@ export const previewOptions = [
 	},
 	{
 		value: 'preview-2',
-		name: 'Out of sotck',
```
S O T C K !

sunyatasattva commented on assets/js/blocks/stock-filter/utils.ts:
```diff
@@ -28,6 +28,18 @@ export const getActiveFilters = (
+export const formatSlug = ( slug: string ) =>
+	slug
+		.trim()
+		.replace( /\s/g, '' )
+		.replace( /_/g, '-' )
+		.replace( /-+/g, '-' )
+		.replace( /[^a-zA-Z0-9-]/g, '' );
```
You should be able to replace those 4 calls with just two, right? Or am I missing something?

kmanijak replied:
Right, `.replace( /\s/g, '' ) can be merged with `.replace( /[^a-zA-Z0-9-]/g, '' );`.

About the `.replace( /_/g, '-' ).replace( /-+/g, '-' )` is not equal to `.replace( /_|-+/g, '-' )` unless you had something else in mind? In case of two following operations the sequence of dashes and underscores will be shrinked to single character, in the latter it won't. 🤔 But regex is not my strongest field, I may totally be missing something 😅

sunyatasattva commented on tests/e2e/specs/backend/stock-filter.test.js:
```diff
@@ -49,5 +49,26 @@ describe( `${ block.name } Block`, () => {
+			// Turn the display style to Dropdown
+			await expect( page ).toClick( 'button', { text: 'Dropdown' } );
+
+			await page.waitForSelector(
+				'.wc-block-stock-filter.style-dropdown'
+			);
+			await expect( page ).toMatchElement(
+				'.wc-block-stock-filter.style-dropdown'
+			);
```
Doesn't `toMatchElement` automatically wait for selector also? 🤔

sunyatasattva reviewed (changes requested):
Really nice UI improvement, well done! Works nice! I've got a few code comments :)

kmanijak committed:
Change the name of the filter blocks in the test files

kmanijak committed:
Remove unnecessary waiting step in E2E test for Filter by Stock

toMatchElement waits for an element for 30s by itself, hence waitForSelector usage was removed

kmanijak committed:
Improve the STOCK_STATUS_OPTIONS type handling

kmanijak committed:
Rename vague  variable to

kmanijak committed:
Remove unnecessary type

kmanijak committed:
Remove unnecessary style

kmanijak committed:
Extract onDropdownChange function instead of inline arrow function

kmanijak commented:
>Could you clarify which limitations you mean on the FormTokenField? I've used the component from Gutenberg in the exact same scenario and it works fine in my use-case:

@sunyatasattva , hmm.. I checked your implementation in [PQ inspector-controls](https://github.com/woocommerce/woocommerce-blocks/blob/590d5945f930775a1638ba2d5453d5afbf98a610/assets/js/blocks/product-query/inspector-controls.tsx) and it's much simpler. I was basing on the Filter by Attribute assuming there are cases covered I'm not aware of. Let me simplify the implementation here before next round of review.

kmanijak commented:
Followup to the previous comment: I tried to simplify the usage, but then I realised I cannot achieve the same result as yourself. It's because in the Filter by Stock we use `FormTokenField` from `wordpress-components` (which points to `@wordpress-components: 14.2.0`), while in Editor you used `@wordpress-components: 19.1.5`. I did a quick check and replaced `FormTokenField` in PQ inspector-controls to the old one and I can reproduce the problem: using whitespace in the input shows no suggestions (see the video below). I encountered the same in Filter by Stock and by playing with props I was able to achieve the result with the limitations as described.

https://user-images.githubusercontent.com/20098064/206665801-f01ad6ce-0d9c-4d91-acd7-113394e20416.mov

My suggestion would be to leave it as it is and improve it once we update the `wordpress-components`, what do you think?

Having said that, we can go with second round of review:
- comments with 👍 reaction were addressed
- others have a response.

PS. That's a lot for a thorough review 🙏

gigitux commented:
My review is only related to the feature. I didn't check the code.
Everything works great. I noticed a strange behavior with the order of the labels:

https://user-images.githubusercontent.com/4463174/207090261-d20c13ec-7854-48ce-b628-eb67a8adb4d0.mp4

For me, it isn't a blocker, so I suggest creating an issue, and we will address the fix in the next iteration 👍

kmanijak committed:
Merge branch 'trunk' into add/filter-by-stock-dropdown

kmanijak committed:
Fix overlaping dropdown content with the wrapper when Filter by Stock was set to single

The fix was provided with another PR #7771 and had to be resued here

nefeline reviewed (approved):
Great Job 🎉 ! I've tested multiple scenarios, and they all worked as expected:

### Display Style: Dropdown - Single

✅  Confirmed dropdown has stock status options to choose from.
✅  Confirmed when choosing one option, the page refreshes, and the filter is applied.
✅  Confirmed that when choosing another option, the previous filter is replaced by the new one, the page refreshes, and the filter is applied.
✅ Confirmed both the products and the selected filters are correctly displayed when changing between filters.

### Display Style: Dropdown - Multiple

✅ Confirmed dropdown has stock status options to choose from.
✅ Confirmed that when choosing a single option page refreshes & the filter is applied.
✅ Confirmed that when choosing a second option, both filters are in place.
✅ Confirmed both the products and the selected filters are correctly displayed when changing between filters.

### Display Style: List - (Single and Multiple)

✅ Confirmed that when changing the filter settings to list, everything works as expected, for both: single and multiple

### Apply button

✅ Confirmed the filters are not applied until the "Apply" button is clicked.

🗒️  I've noticed a small spacing issue when using the Stock Filter (with the Apply button enabled) alongside the Products Block:

<img width="1110" alt="Screenshot 2022-12-13 at 14 49 58" src="https://user-images.githubusercontent.com/15730971/207355405-ebf22ef8-cc58-469a-9c72-81bccd1a48d2.png">

This is reproducible on trunk as well (pre-existing), so not a blocker here.

✅ Tested other scenarios, including filtering in conjunction with other filters (by rating, by attribute, etc.): confirmed things are working as expected.

kmanijak committed:
Merge branch 'trunk' into add/filter-by-stock-dropdown

nefeline commented:
Minor: I've just noticed that in the editor, the color of this block differs from others:

<img width="839" alt="Screenshot 2022-12-13 at 15 52 22" src="https://user-images.githubusercontent.com/15730971/207367080-525d5686-0ddf-4abd-8be0-58f32140b8f8.png">

What do you think about aiming for consistency and ensuring the same initial default color is used here?

gigitux commented:
@nefeline, what theme do you have? With TT3, I'm not able to replicate it. 

![image](https://user-images.githubusercontent.com/4463174/207390576-7ded8a55-5fe1-47dc-a765-f80a75567336.png)

I don't consider it a blocker (not sure that this PR introduces the issue). Does it make sense to create a new issue?

nefeline commented:
Hi @gigitux ! I don't consider this a blocker, either. @kmanijak and I were already chatting about creating a separate issue earlier today: the issue happens when you configure the color via global settings.

kmanijak committed:
Merge branch 'trunk' into add/filter-by-stock-dropdown

kmanijak committed:
Merge branch 'trunk' into add/filter-by-stock-dropdown

kmanijak commented:
@nefeline, good catch and thanks for raising it! 🙌 

I created an issue for that https://github.com/woocommerce/woocommerce-blocks/issues/7940. This will be easier to fix for all three filters that support dropdowns at once: by Attribute, by Rating and by Stock once this PR is merged.
